src/pm/hydra/proxy/pmip_pg.c

```diff
@@ -264,3 +264,45 @@ void PMIP_bcast_signal(int sig)
         }
     }
 }
+
+int PMIP_pg_local_to_global_id(struct pmip_pg *pg, int local_id)
+{
+    int rem1, rem2, layer, ret;
+
+    if (local_id < pg->global_core_map.local_filler)
+        ret = pg->pmi_id_map.filler_start + local_id;
+    else {
+        rem1 = local_id - pg->global_core_map.local_filler;
+        layer = rem1 / pg->global_core_map.local_count;
+        rem2 = rem1 - (layer * pg->global_core_map.local_count);
+
+        ret = pg->pmi_id_map.non_filler_start + (layer * pg->global_core_map.global_count) + rem2;
+    }
+
+    return ret;
+}
+
+int PMIP_pg_global_to_local_id(struct pmip_pg *pg, int global_id)
+{
+    if (global_id < pg->pmi_id_map.filler_start) {
+        return -1;
+    } else if (global_id < pg->pmi_id_map.non_filler_start) {
+        int rem1 = global_id - pg->pmi_id_map.filler_start;
+        if (rem1 < pg->global_core_map.local_filler) {
+            return rem1;
+        } else {
+            return -1;
+        }
+    } else {
+        int rem1 = global_id - pg->pmi_id_map.non_filler_start;
+        int layer = rem1 / pg->global_core_map.global_count;
+        int rem2 = rem1 - layer * pg->global_core_map.global_count;
+
+        if (rem2 < pg->global_core_map.local_count) {
+            return pg->global_core_map.local_filler + (layer * pg->global_core_map.local_count)
+                + rem2;
+        } else {
+            return -1;
+        }
+    }
+}
```

Comment:
i trust these are right, but it sure would be nice to have a high-level explanation of what is going on. the logic and naming is tough to decipher.

Reply:
![image](https://github.com/user-attachments/assets/690ea938-4201-4541-acf5-0d045d120500)

Does this help?

Reply:
I updated the diagram showing the layer count. Yeah, layer is the wrap around. It assumes a fully-filled layer so `local_id += layer * local_count` and `global_id += layer * global_count`.

Reply:
maybe more importantly, what is the "filler"? searching the code it seems like extra processes launched that are not part of the initial world? is that right?

Reply:
Imagine a spawned new world. The original world doesn't have fillers - fillers are 0. But the second world need "fill" whatever leftover count from the previous allocation first. That's what filler refers to.

Reply:
ok. so the global ids are independent of the world they are in? for instance rank 0 would have non-zero local and global id values?

Reply:
No. Global ids are world ranks. It always start 0. So the `filler_start` at the first node is always `0`. It will be greater than 0 on later nodes if the previous nodes have non-0 `local_filler`.

Reply:
I'm still pretty lost but I guess we can discuss offline.